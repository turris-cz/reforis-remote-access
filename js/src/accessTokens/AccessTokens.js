/*
 * Copyright (C) 2019-2021 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect, useState } from "react";

import { useAPIGet, withErrorMessage, withSpinnerOnSending } from "foris";
import PropTypes from "prop-types";

import API_URLs from "API";

import CreateTokenForm from "./CreateTokenForm";
import Tokens from "./Tokens";

AccessTokens.propTypes = {
    ws: PropTypes.object.isRequired,
};

export default function AccessTokens({ ws }) {
    const [authority, getAuthority] = useAPIGet(API_URLs.authority);
    useEffect(() => {
        getAuthority();
    }, [getAuthority]);

    return (
        <>
            <h1>{_("Access Tokens")}</h1>
            <TokensWrapperWithErrorAndSpinner
                apiState={authority.state}
                ws={ws}
                authority={authority.data}
            />
        </>
    );
}

const TokensWrapperWithErrorAndSpinner = withErrorMessage(
    withSpinnerOnSending(TokensWrapper)
);

TokensWrapper.propTypes = {
    ws: PropTypes.object.isRequired,
    authority: PropTypes.object.isRequired,
};

function TokensWrapper({ ws, authority }) {
    const [generating, setGenerating] = useState(false);

    if (authority.status !== "ready") {
        return (
            <p>
                {_(
                    "You need to generate certificate authority to create tokens."
                )}
            </p>
        );
    }

    return (
        <>
            <p>
                {_(
                    `You need to generate a token file for each client you want to grant access to.`
                )}
            </p>
            <p>
                {_(
                    `To use the token with another Turris router you need to transfer it using the "Remote Devices" plugin on the client device.`
                )}
            </p>
            <CreateTokenForm
                generating={generating}
                setGenerating={setGenerating}
            />
            <Tokens ws={ws} setGenerating={setGenerating} />
        </>
    );
}
