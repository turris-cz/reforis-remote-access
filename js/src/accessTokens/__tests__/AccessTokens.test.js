/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import {
    render,
    wait,
    act,
    getByText,
    getByLabelText,
    getByRole,
    queryByRole,
    fireEvent,
} from "foris/testUtils/customTestRender";
import { mockJSONError } from "foris/testUtils/network";
import { WebSockets } from "foris";
import mockAxios from "jest-mock-axios";

import AccessTokens from "../AccessTokens";

describe("<AccessTokens />", () => {
    let webSockets;
    let container;

    function getAddButton() {
        return getByText(container, "Add");
    }

    function submitTokenForm(name) {
        const tokenNameInput = getByLabelText(container, "Token name");
        fireEvent.change(tokenNameInput, { target: { value: name } });
        const addButton = getAddButton();
        fireEvent.click(addButton);
    }

    beforeEach(() => {
        webSockets = new WebSockets();
        ({ container } = render(<AccessTokens ws={webSockets} />));
    });

    it("should render spinner", () => {
        expect(container).toMatchSnapshot();
    });

    it("should render error", async () => {
        expect(mockAxios.get).toBeCalledWith(
            "/reforis/remote-access/api/authority",
            expect.anything()
        );
        mockJSONError();
        await wait(() => expect(container).toMatchSnapshot());
    });

    it("should handle invalid CA", async () => {
        mockAxios.mockResponse({ data: { status: "whatever" } });
        await wait(() =>
            getByText(
                container,
                "You need to generate certificate authority in order to create tokens."
            )
        );
    });

    it("should handle proper CA", async () => {
        mockAxios.mockResponse({ data: { status: "ready" } });
        await wait(() => expect(container).toMatchSnapshot());
    });

    it("should add new token to the table", async () => {
        // Response to GET authority
        mockAxios.mockResponse({ data: { status: "ready" } });
        await wait(() => getByText(container, "Create new token"));
        // Response to GET tokens
        mockAxios.mockResponse({ data: [] });
        await wait(() =>
            getByText(container, "There are no tokens added yet.")
        );

        const name = "another_turris";
        submitTokenForm(name);
        expect(getAddButton().disabled).toBe(true);
        // Handle request
        expect(mockAxios.post).toBeCalledWith(
            "/reforis/remote-access/api/tokens",
            { name },
            expect.anything()
        );
        mockAxios.mockResponse({ data: { task_id: "1234" } });
        await wait(() => getByText(container, "Create new token"));

        // Generating new token
        act(() =>
            webSockets.dispatch({
                module: "remote",
                action: "generate_token",
                data: { status: "token_generating" },
            })
        );
        // Third GET request is the second one for "tokens" endpoint
        expect(mockAxios.get).toHaveBeenNthCalledWith(
            3,
            "/reforis/remote-access/api/tokens",
            expect.anything()
        );
        mockAxios.mockResponse({
            data: [{ id: "T1", name, status: "generating" }],
        });
        // New token appears (with spinner)
        await wait(() => expect(getByText(container, name)).toBeDefined());
        expect(getByRole(container, "status")).toBeTruthy();

        // Token is ready
        act(() =>
            webSockets.dispatch({
                module: "remote",
                action: "generate_token",
                data: { status: "succeeded" },
            })
        );
        expect(mockAxios.get).toHaveBeenNthCalledWith(
            4,
            "/reforis/remote-access/api/tokens",
            expect.anything()
        );
        mockAxios.mockResponse({
            data: [{ id: "T1", name, status: "succeeded" }],
        });
        // Table is refreshed
        await wait(() => expect(getByText(container, name)).toBeDefined());
        // Spinner is gone
        expect(queryByRole(container, "status")).toBeFalsy();
        // Download is available
        expect(getByText(container, "Download")).toBeTruthy();
        // Add button is enabled again
        expect(getAddButton().disabled).toBe(false);
    });
});
