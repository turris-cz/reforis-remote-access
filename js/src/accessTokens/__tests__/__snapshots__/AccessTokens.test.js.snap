// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AccessTokens /> should handle proper CA 1`] = `
<div>
  <h1>
    Access Tokens
  </h1>
  <p>
    You need to generate a token file for each client you want to grant access to.
  </p>
  <p>
    To use the token with another Turris router you need to transfer it using the "Remote Devices" plugin on the client device.
  </p>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Create Token
    </h2>
    <form>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="1"
        >
          Token name
        </label>
        <div
          class="input-group"
        >
          <input
            class="form-control is-invalid"
            id="1"
            type="text"
            value=""
          />
        </div>
        <div
          class="invalid-feedback"
        >
          Name cannot be empty
        </div>
        <div
          class="form-text"
        >
          <small>
            Shorter than 64 characters. Only alphanumeric characters, dots, dashes and underscores.
          </small>
        </div>
      </div>
      <div
        class="text-end"
      >
        <button
          class="btn btn-primary col-12 col-md-3 col-lg-2 d-inline-flex justify-content-center align-items-center"
          data-testid="add-token-button"
          disabled=""
          type="submit"
        >
          <span>
            Add
          </span>
        </button>
      </div>
    </form>
  </div>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Tokens
    </h2>
    <div
      class="spinner-wrapper my-3 text-center"
    >
      <div
        class="spinner-border"
        role="status"
      >
        <span
          class="sr-only"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<AccessTokens /> should render error 1`] = `
<div>
  <h1>
    Access Tokens
  </h1>
  <p
    class="text-center text-danger"
  >
    An error occurred while fetching data.
  </p>
</div>
`;

exports[`<AccessTokens /> should render spinner 1`] = `
<div>
  <h1>
    Access Tokens
  </h1>
  <div
    class="spinner-wrapper my-3 text-center"
  >
    <div
      class="spinner-border"
      role="status"
    >
      <span
        class="sr-only"
      />
    </div>
  </div>
</div>
`;
